### DISCUSSION TITLE
How to dynamically generate static JSON pages/API?

### DISCUSSION THREAD

**Discussion by inwardmovement** [vercel/next.js#38704] [Help]
Say I have 3 posts, and I want a dynamically generated static `/api/posts.json` (or `/api-static/posts.json`) page to output:

```
{
  "postsCount": "3"
}
```

Where `3` would be the number of posts, computed at build time.

How can I achieve that?
Shouldn't Next.js provide an easy/built-in way to build static APIs, and allow `/api` to be prerendered (and properly document all that)?

**Comment by GabenGar** [comment]
If it's static just use the data in pages directly.

**Reply by inwardmovement** [reply]
But will the output "format" be `application/json` ?
Don't we need a hack like [this](https://www.grouparoo.com/blog/dynamic-json-pages-with-nextjs#method-2-static-file)?

Can you provide a simple example please?

**Reply by GabenGar** [reply]
Why do you need static API pages in the first place? If for use in your pages, then just consume the data directly in them at build time. If it comes from somewhere, then there is no point duplicating the data.

**Reply by inwardmovement** [reply]
I have website A that statically builds a JSON endpoint to expose local static data, and want to get this data to build website B. Think of it as `pages/api/` but prerendered at build time.

With Hugo you would use [custom output formats](https://forestry.io/blog/build-a-json-api-with-hugo/).

How to achieve that with Next? (in a "native" way
 I mean. Because I can work around this with a custom pre-build script that would build `public/data.json` but it's not what I want).

**Reply by GabenGar** [reply]
Why not just use the data in website B directly?

**Reply by inwardmovement** [reply]
> Why not just use the data in website B directly?

Because the data comes from website A. It exposes local static data. Ex:
I have 3 posts on website A, and I want `websiteA.com/data.json` to expose the number of posts (3), so that websiteB can fetch `websiteA.com/data.json` to use the number of websiteA's posts in its build.

**Reply by GabenGar** [reply]
And what prevents website B to be fed with json directly? It can be done as part of  build pipeline, no need to add a volatile chain of static http requests to it.

**Reply by inwardmovement** [reply]
So how do I pass the number of WebsiteA posts to WebsiteB build?

WebsiteA:
- post-1.md
- post-2.md
- posts.json > outputs number of posts (2) at `/data.json`.

WebsiteB build:
fetch `websiteA/data.json` > `websiteA.postsNumber == 2`

This is what I do with Hugo. The question is not why, but how to do it with Next.

**Reply by Noitidart** [reply]
@inwardmovement were you able to figure it out. I was also trying to create ISG JSON pages - found this discussion on how to make non-static JSON pages - but for that we might as well use an API route - https://github.com/vercel/next.js/discussions/14991

**Reply by inwardmovement** [reply]
No I didn't, I'm still hoping for help. It would suck to have to use SSR to generate static pages...

**Reply by Noitidart** [reply]
Darn, thanks. I'll update you if I figure it out too.

**Comment by pirtlj** [comment]
It would be nice to leverage static regeneration with json files.  This would be great for metadata that changes slowly which could be regenerated and cached on demand.
